File: server/utils/textFormatter.js
// Basic logger
const logger = { info: console.log, warn: console.warn, error: console.error };

const TextFormatter = {
    formatNovelContent: (content, useTraditionalQuotes = true) => {
        if (!content || typeof content !== 'string') {
            return content;
        }

        let formatted = content;

        // Normalize line endings
        formatted = formatted.replace(/\r\n|\r/g, '\n');

        // Unify quotes
        if (useTraditionalQuotes) {
            // Convert English quotes to Chinese quotes
            // This needs to be careful not to mess up already formatted quotes or nested quotes.
            // A simple regex might be too naive for complex cases.
            // Python version: content = re.sub(r'"([^"]*)"', r'「\1」', content) (multiple times)
            // JS equivalent is trickier due to lack of lookbehinds in all engines, but modern JS supports them.
            // For simplicity, let's do a basic replacement.
            formatted = formatted.replace(/"([^"]*?)"/g, '「$1」');
             // Python version did it 3 times, perhaps for overlapping matches or specific cases.
            formatted = formatted.replace(/"([^"]*?)"/g, '「$1」');
            formatted = formatted.replace(/"([^"]*?)"/g, '「$1」');
        } else {
            // Convert Chinese quotes to English quotes
            formatted = formatted.replace(/「([^」]*?)」/g, '"$1"');
        }

        // Paragraph formatting (Python version was complex)
        // Simplified version: ensure double newlines after sentences ending with specific punctuation.
        // formatted = formatted.replace(/([。！？])([^」\n])/g, '$1\n\n$2'); // Python: ([。！？])([^」
])
        // formatted = formatted.replace(/([」])([。！？])([^」\n])/g, '$1$2\n\n$3'); // Python: ([」])([。！？])([^」
])

        // A simpler approach for JS: Split by lines, process, then join.
        const lines = formatted.split('\n');
        let processedLines = [];
        for (let i = 0; i < lines.length; i++) {
            let line = lines[i].trim();
            if (line) {
                 // Ensure sentences end with appropriate punctuation (very basic)
                if (/[a-zA-Z0-9\u4e00-\u9fff]$/.test(line) && !/[。！？？」]$/.test(line)) {
                    line += '。';
                }
                processedLines.push(line);
            } else if (processedLines.length > 0 && processedLines[processedLines.length -1] !== '') {
                // Keep single empty lines between paragraphs, but not multiple
                processedLines.push('');
            }
        }

        // Remove leading/trailing empty lines and ensure max one empty line between content lines
        formatted = processedLines.filter((line, index, arr) => {
            return line !== '' || (index > 0 && arr[index-1] !== '');
        }).join('\n');

        // Clean up multiple newlines that might have been introduced or were already there
        formatted = formatted.replace(/\n{3,}/g, '\n\n');

        return formatted.trim();
    }
    // _formatParagraphs, _formatDialogue, _fixPunctuation from Python could be translated here if needed
    // For now, keeping it simpler.
};

module.exports = { TextFormatter };

File: novel_writer/services/text_formatter.py
"""
文本格式化服務模組
負責小說內容的格式化處理，包括引號、段落、對話和標點符號的統一處理。
"""

import re
from typing import Optional


class TextFormatter:
    """文本格式化器"""
    
    @staticmethod
    def format_novel_content(content: str, use_traditional_quotes: bool = True) -> str:
        """格式化小說內容"""
        if not content:
            return content
        
        # 統一引號
        if use_traditional_quotes:
            # 將所有英文引號轉換為中文引號
            content = re.sub(r'"([^"]*)"', r'「\1」', content)
            content = re.sub(r'"([^"]*)"', r'「\1」', content)
            content = re.sub(r'"([^"]*)"', r'「\1」', content)
        else:
            # 將所有中文引號轉換為英文引號
            content = re.sub(r'「([^」]*)」', r'"\1"', content)
        
        # 處理段落分行
        content = TextFormatter._format_paragraphs(content)
        
        # 處理對話格式
        content = TextFormatter._format_dialogue(content, use_traditional_quotes)
        
        # 清理多餘的空行
        content = re.sub(r'\n{3,}', '\n\n', content)
        
        # 確保句子結尾有適當的標點
        content = TextFormatter._fix_punctuation(content)
        
        return content.strip()
    
    @staticmethod
    def _format_paragraphs(content: str) -> str:
        """格式化段落分行"""
        # 在句號、感嘆號、問號後添加換行（如果後面不是換行的話）
        content = re.sub(r'([。！？])([^」\n])', r'\1\n\n\2', content)
        
        # 在引號結束後如果有句號等，也要換行
        content = re.sub(r'([」])([。！？])([^」\n])', r'\1\2\n\n\3', content)
        
        # 處理對話後的描述
        content = re.sub(r'([」])([^。！？\n][^」]*?[。！？])', r'\1\n\n\2', content)
        
        return content
    
    @staticmethod
    def _format_dialogue(content: str, use_traditional_quotes: bool) -> str:
        """格式化對話"""
        if use_traditional_quotes:
            # 確保對話前有適當的分行
            content = re.sub(r'([。！？])(\s*)([^」\n]*?)「', r'\1\n\n\3「', content)
        else:
            # 確保對話前有適當的分行
            content = re.sub(r'([。！？])(\s*)([^"\n]*?)"', r'\1\n\n\3"', content)
        
        return content
    
    @staticmethod
    def _fix_punctuation(content: str) -> str:
        """修復標點符號"""
        # 確保句子結尾有標點
        lines = content.split('\n')
        fixed_lines = []
        
        for line in lines:
            line = line.strip()
            if line and not re.search(r'[。！？」"]$', line):
                # 如果行末沒有標點，添加句號
                if re.search(r'[a-zA-Z0-9\u4e00-\u9fff]$', line):
                    line += '。'
            fixed_lines.append(line)
        
        return '\n'.join(fixed_lines)